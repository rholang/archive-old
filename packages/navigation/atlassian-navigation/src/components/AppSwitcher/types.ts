import { IconButtonProps } from '../IconButton/types';

export type AppSwitcherProps = Omit<IconButtonProps, 'icon' | 'tooltip'> & {
  tooltip: string;
};
