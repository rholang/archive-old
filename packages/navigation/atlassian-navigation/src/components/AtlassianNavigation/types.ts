import React, { ReactNodeArray, ReactNode } from 'react';

import { NavigationTheme } from '../../theme';

export type AtlassianNavigationProps = {
  primaryItems: ReactNodeArray;
  renderAppSwitcher?: React.ComponentType<{}>;
  renderCreate?: React.ComponentType<{}>;
  renderHelp?: React.ComponentType<{}>;
  renderNotifications?: React.ComponentType<{}>;
  renderProductHome: React.ComponentType<{}>;
  renderProfile?: React.ComponentType<{}>;
  renderSearch?: React.ComponentType<{}>;
  renderSignIn?: React.ComponentType<{}>;
  renderSettings?: React.ComponentType<{}>;
  moreLabel?: ReactNode;
  theme?: NavigationTheme;
};

export type NavigationSkeletonProps = {
  primaryItemsCount?: number;
  secondaryItemsCount?: number;
  theme?: NavigationTheme;
};
