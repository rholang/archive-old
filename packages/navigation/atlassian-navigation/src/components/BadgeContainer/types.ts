import { ComponentType, ReactNode } from 'react';

export type BadgeProps = {
  badge: ComponentType<{}>;
  children: ReactNode;
};
