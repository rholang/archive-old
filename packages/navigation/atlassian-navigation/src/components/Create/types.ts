export type CreateProps = {
  text: string;
  onClick: (e: React.MouseEvent<HTMLElement>) => void;
};
