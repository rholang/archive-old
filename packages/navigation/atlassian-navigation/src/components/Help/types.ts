import { IconButtonProps } from '../IconButton/types';

export type HelpProps = Omit<IconButtonProps, 'icon' | 'tooltip'> & {
  tooltip: string;
};
