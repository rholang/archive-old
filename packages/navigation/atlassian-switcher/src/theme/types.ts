export type Appearance = 'drawer' | 'standalone';

export type ThemingPublicApi = {
  primaryTextColor?: string;
  secondaryTextColor?: string;
  primaryHoverBackgroundColor?: string;
  secondaryHoverBackgroundColor?: string;
};

export type ItemStateTokens = {
  background?: string;
  text?: string;
  secondaryText?: string;
};

export type ThemeTokens = {
  padding?: any;
  hover?: ItemStateTokens;
  default?: ItemStateTokens;
  active?: ItemStateTokens;
};

export type ThemeProps = {};

export type ApplyThemeFn = (props: ThemeProps) => ThemeTokens;

type ThemeFn = (theme: ApplyThemeFn, props: ThemeProps) => ThemeTokens;

export type CustomThemeResult = {
  itemTheme: ThemeFn;
  childItemTheme: ThemeFn;
  topLevelItemWrapperTheme: ThemeFn;
};

export type WithTheme = {
  theme?: ThemingPublicApi;
  appearance?: Appearance;
};

export type Themed<T> = T & { theme?: ThemeFn; appearance?: Appearance };
