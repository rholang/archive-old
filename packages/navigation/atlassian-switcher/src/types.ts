import { UIAnalyticsEvent } from '@atlaskit/analytics-next';
import { WithTheme } from './theme/types';
import { AvailableProductsDataProvider } from './providers/products-data-provider';

export interface TriggerXFlowCallback {
  (
    productKey: string,
    sourceComponent: string,
    event: any,
    analyticsEvent: UIAnalyticsEvent,
  ): void;
}
export interface DiscoverMoreCallback {
  (event: any, analyticsEvent: UIAnalyticsEvent): void;
}

export interface WithCloudId {
  cloudId: string;
}

export enum RecentContainerType {
  JIRA_PROJECT = 'jira-project',
  CONFLUENCE_SPACE = 'confluence-space',
}

export interface RecentContainer {
  name: string;
  url: string;
  objectId: string;
  iconUrl: string;
  type: RecentContainerType;
}

export interface CustomLink {
  key: string;
  label: string;
  link: string;
  local: boolean;
}

export enum Permissions {
  MANAGE = 'manage',
  CAN_INVITE_USERS = 'invite-users',
  ADD_PRODUCTS = 'add-products',
}

export enum Product {
  BITBUCKET = 'bitbucket',
  CONFLUENCE = 'confluence',
  HOME = 'home',
  JIRA = 'jira',
  SITE_ADMIN = 'site-admin',
  TRUSTED_ADMIN = 'trusted-admin',
}

export enum Feature {
  disableCustomLinks = 'disableCustomLinks',
  disableRecentContainers = 'disableRecentContainers',
  disableHeadings = 'disableHeadings',
  xflow = 'xflow',
  isDiscoverMoreForEveryoneEnabled = 'isDiscoverMoreForEveryoneEnabled',
  // EMCEE stands for Embedded Marketplace with in the product
  isEmceeLinkEnabled = 'isEmceeLinkEnabled',
  // Enable Discover section - group suggested product links in Discover section
  isDiscoverSectionEnabled = 'isDiscoverSectionEnabled',
}

export type FeatureFlagProps = {
  // Custom links are enabled by default for Jira and Confluence, this feature flag allows to hide them. Custom links are not supported by the switcher in any other products.
  disableCustomLinks?: boolean;
  // Hide recent containers. Recent containers are enabled by default.
  disableRecentContainers?: boolean;
  // Remove section headers - useful if something else is providing them. i.e: trello inline dialog.
  disableHeadings?: boolean;
  // Enable discover more.
  isDiscoverMoreForEveryoneEnabled?: boolean;
  // Enable Embedded Marketplace within the product.
  isEmceeLinkEnabled?: boolean;
  // Enable Discover section - group suggested product links in Discover section
  isDiscoverSectionEnabled?: boolean;
};

export type FeatureMap = { [key in Feature]: boolean };

export type CustomLinksResponse = CustomLink[];

export type ProvisionedProducts = { [key in WorklensProductType]?: boolean };

export interface CurrentSite {
  url: string;
  products: AvailableProduct[];
}

export interface UserSiteDataResponse {
  currentSite: CurrentSite;
  provisionedProducts: ProvisionedProducts;
}

export interface XFlowSettingsResponse {
  'product-suggestions-enabled'?: boolean;
}

export interface UserPermissionResponse {
  permitted: boolean;
}

export interface RecentContainersResponse {
  data: Array<RecentContainer>;
}

export enum WorklensProductType {
  JIRA_BUSINESS = 'JIRA_BUSINESS',
  JIRA_SERVICE_DESK = 'JIRA_SERVICE_DESK',
  JIRA_SOFTWARE = 'JIRA_SOFTWARE',
  CONFLUENCE = 'CONFLUENCE',
  OPSGENIE = 'OPSGENIE',
  BITBUCKET = 'BITBUCKET',
  STATUSPAGE = 'STATUSPAGE',
  TRELLO = 'TRELLO',
}

export type AvailableProduct =
  | {
      activityCount: number;
      productType:
        | WorklensProductType.JIRA_BUSINESS
        | WorklensProductType.JIRA_SERVICE_DESK
        | WorklensProductType.JIRA_SOFTWARE
        | WorklensProductType.CONFLUENCE;
    }
  | AvailableProductWithUrl;

interface AvailableProductWithUrl {
  activityCount: number;
  productType:
    | WorklensProductType.BITBUCKET
    | WorklensProductType.OPSGENIE
    | WorklensProductType.STATUSPAGE // assuming that the URL is provided by TCS (same as Opsgenie)
    | WorklensProductType.TRELLO;
  url: string;
}

export interface AvailableSite {
  adminAccess: boolean;
  availableProducts: AvailableProduct[];
  avatar: string | null;
  cloudId: string;
  displayName: string;
  url: string;
}

export interface AvailableProductsResponse {
  sites: AvailableSite[];
}

export enum ProductKey {
  CONFLUENCE = 'confluence.ondemand',
  JIRA_CORE = 'jira-core.ondemand',
  JIRA_SOFTWARE = 'jira-software.ondemand',
  JIRA_SERVICE_DESK = 'jira-servicedesk.ondemand',
  OPSGENIE = 'opsgenie',
}

export type RecommendationsEngineResponse = RecommendationItem[];

export interface RecommendationItem {
  productKey: ProductKey;
}

export type RecommendationsFeatureFlags = {
  [key: string]: string | boolean;
};

export interface SwitcherChildItem {
  href: string;
  label: string;
  avatar: string | null;
}

export type AtlassianSwitcherProps = WithTheme & {
  // Product name used for analytics events
  product: string;
  // Optional cloudID, should be provided for tenanted applications.
  cloudId?: string;
  // Optional callback to be exectuted after an XFlow event is triggered.
  triggerXFlow?: TriggerXFlowCallback;
  // Optional callback to be exectuted after a user clicks on discover more.
  onDiscoverMoreClicked?: DiscoverMoreCallback;
  // A map of feature flags used by the XFlow recommendations engine.
  recommendationsFeatureFlags?: RecommendationsFeatureFlags;
  // Optional custom provider for available products
  availableProductsDataProvider?: AvailableProductsDataProvider;
} & FeatureFlagProps;
