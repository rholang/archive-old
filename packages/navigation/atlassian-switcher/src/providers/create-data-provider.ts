import asDataProvider, { DataProviderProps } from './as-data-provider';
import { fetchJson } from '../utils/fetch';
import { withCached, WithCached } from '../utils/with-cached';

/**
 * Some items might be using the type `ExportedDataProvider` instead due to errors with
 * the generated documentation
 */
export type DataProvider<T> = {
  fetchMethod: WithCached<(param: object) => Promise<T>>;
  ProviderComponent: DataProviderProps<T> & any;
};

/**
 * `WithCached` within `DataProvider` breaks the documentation with error:
 * `Error: Missing converter for: TSMappedType` due to usage of Proxy function
 * so we are exporting a simpler type here just for the docs. There has been reported
 * on their repo already: https://github.com/atlassian/extract-react-types/issues/75
 */
export type ExportedDataProvider<T> = {
  fetchMethod: (param: object) => Promise<T>;
  ProviderComponent: DataProviderProps<T> & any;
};

export const createProvider = <T>(
  name: string,
  url: string,
): DataProvider<T> => {
  const fetchMethod = withCached((param: object) => fetchJson<T>(url));

  return {
    fetchMethod,
    ProviderComponent: asDataProvider(name, fetchMethod, fetchMethod.cached),
  };
};
