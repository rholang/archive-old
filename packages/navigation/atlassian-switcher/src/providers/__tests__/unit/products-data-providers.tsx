import * as React from 'react';
import { shallow } from 'enzyme';

import {
  AvailableProductsProvider,
  prefetchAvailableProducts,
} from '../../products-data-provider';

import { AvailableProductsResponse } from '../../../types';

import { createProvider } from '../../create-data-provider';

describe('products-data-providers', () => {
  test('should render using the default provider', () => {
    const wrapper = shallow(
      <AvailableProductsProvider>{items => items}</AvailableProductsProvider>,
    );
    expect(wrapper).toMatchSnapshot();
  });

  test('should render using a custom provider', () => {
    const customProvider = createProvider<AvailableProductsResponse>(
      'my-provider',
      'my-new-endpoint',
    );
    const wrapper = shallow(
      <AvailableProductsProvider availableProductsDataProvider={customProvider}>
        {items => items}
      </AvailableProductsProvider>,
    );
    expect(wrapper).toMatchSnapshot();
  });

  test('should prefetch using the custom provider if passed down', () => {
    const fetchMethod = jest.fn();

    prefetchAvailableProducts({
      fetchMethod,
    } as any);

    expect(fetchMethod).toBeCalledTimes(1);
  });
});
