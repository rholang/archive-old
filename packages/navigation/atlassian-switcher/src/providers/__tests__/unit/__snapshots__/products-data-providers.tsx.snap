// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`products-data-providers should render using a custom provider 1`] = `
<WithAnalyticsEvents(DataProvider(my-provider))>
  <Component />
</WithAnalyticsEvents(DataProvider(my-provider))>
`;

exports[`products-data-providers should render using the default provider 1`] = `
<WithAnalyticsEvents(DataProvider(availableProducts))>
  <Component />
</WithAnalyticsEvents(DataProvider(availableProducts))>
`;
